extensions/wikia/CategoryPagination/CategoryPaginationViewer.class.php

```diff
@@ -2,102 +2,221 @@
 	 */
-	private function getActualFrom( $type, $page ) {
-		$dbr = wfGetDB( DB_SLAVE, 'category' );
-
-		// if beyond the last page, show the last page instead
-		$totalPages = ceil( $this->counts[$type] / $this->limit );
-		if ( $page > $totalPages && $totalPages >= 1 ) {
-			$page = $totalPages;
-		}
-
-		$conditions = [
-			'cl_to' => $this->title->getDBkey(),
+	private function processSection( string $type, callable $processRow, callable $generateContent ) {
+		$conds = [
 			'cl_type' => $type,
+			'cl_to' => $this->title->getDBkey(),
 		];
 
-		$res = $dbr->select(
+		// Wall uses this hook to filter out wall messages from the category pages
+		wfRunHooks( 'CategoryViewer::beforeCategoryData', array( &$conds ) );
+
+		$dbr = wfGetDB( DB_SLAVE, 'category' );
+		$result = $dbr->select(
 			[ 'page', 'categorylinks' ],
-			[ 'page.*', 'cl_sortkey_prefix' ],
-			$conditions,
+			[ 'page_namespace', 'page_title', 'page_len', 'page_is_redirect', 'cl_sortkey_prefix' ],
+			$conds,
 			__METHOD__,
 			[
-				'USE INDEX' => array( 'categorylinks' => 'cl_sortkey' ),
-				'LIMIT' => 1,
-				'OFFSET' => ( $page - 1 ) * $this->limit,
 				'ORDER BY' => 'cl_sortkey',
+				'LIMIT' => $this->limit,
+				'OFFSET' => $this->limit * ( $this->paginators[$type]->getActivePage() - 1 ),
+				'USE INDEX' => array( 'categorylinks' => 'cl_sortkey' ),
 			],
 			[ 'categorylinks' => [ 'INNER JOIN', 'cl_from = page_id' ] ]
 		);
-		$row = $res->fetchObject();
 
-		if ( $row ) {
-			$title = Title::newFromRow( $row );
-			$humanSortKey = $title->getCategorySortkey( $row->cl_sortkey_prefix );
-			return $humanSortKey;
+		$numberShown = 0;
+		if ( $result->numRows() ) {
+			while ( $row = $result->fetchObject() ) {
+				$title = Title::makeTitle( $row->page_namespace, $row->page_title );
+				$humanSortkey = $title->getCategorySortkey( $row->cl_sortkey_prefix );
+				if ( $processRow( $title, $humanSortkey, $row ) ) {
+					$numberShown += 1;
+				}
+			}
 		}
+		$this->numberShown[$type] = $numberShown;
 
-		return null;
+		if ( $numberShown > 0 ) {
+			$this->content[$type] = $generateContent();
+		}
 	}
 
 	/**
-	 * Get the paginator object for given section
-	 * @param string $type: page, subcat or file
-	 * @return Paginator
+	 * Get the rel="prev/next" links to put to the head
+	 * From all sections, the one with the most pages is selected to decide if there's
+	 * the next page or not
+	 *
+	 * @return string
 	 */
-	public function getPaginator( $type ) {
-		return $this->paginators[$type];
+	private function getPaginationHeadItem() {
+		$numPages = -1;
+		$maxPaginator = null;
+		foreach ( $this->paginators as $paginator ) {
+			if ( $paginator->getPagesCount() > $numPages ) {
+				$numPages = $paginator->getPagesCount();
+				$maxPaginator = $paginator;
+			}
+		}
+		if ( $maxPaginator ) {
+			// TODO: make sure to link without the #mw-pages, #mw-category-media and #mw-subcategories fragments
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this "TODO" comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1135)